File: src/main/scala/com/builtbroken/lib/transform/region/Sphere.scala
package com.builtbroken.lib.transform.region

import java.util

import io.netty.buffer.ByteBuf
import net.minecraft.entity.Entity
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.MathHelper
import net.minecraft.world.World
import com.builtbroken.lib.transform.vector.{IVector3, Vector3}
import com.builtbroken.lib.wrapper.ByteBufWrapper._

/** 3D Ball shaped region in the world.
  * Can be used for collision boxes, and entity detection
  *
  * Created by robert on 12/18/2014.
  */
class Sphere(c: IVector3, var r: Double) extends Shape3D[Sphere](c)
{
  override def set(other: Sphere): Sphere =
  {
    super.set(other)
    this.r = other.r
    return this
  }

  override def +(amount: Double): Sphere = new Sphere(center, r + amount)

  override def +(amount: Sphere): Sphere = new Sphere(new Vector3(center).midPoint(amount.center), r + amount.r)

  override def *(amount: Double): Sphere = new Sphere(center, r * amount)

  override def *(amount: Sphere): Sphere = new Sphere(new Vector3(center).midPoint(amount.center), r * amount.r)

  override def writeByteBuf(data: ByteBuf): ByteBuf =
  {
    super.writeByteBuf(data)
    data <<< r
    return data
  }

  override def writeNBT(nbt: NBTTagCompound): NBTTagCompound =
  {
    super.writeNBT(nbt)
    nbt.setDouble("radius", r)
    return nbt
  }

  override def getSizeX: Double = r * 2

  override def getSizeY: Double = r * 2

  override def getSizeZ: Double = r * 2

  override def getArea: Double = 4 * Math.PI * (r * r)

  override def getVolume: Double = (4 * Math.PI * (r * r * r)) / 3

  override def isWithin(x: Double, y: Double, z: Double): Boolean = new Vector3(x, y, z).subtract(x, y, z).magnitude <= this.r

  def getEntities[E <: Entity](world: World, clazz: Class[E]): java.util.List[E] =
  {
    val list: java.util.List[E] = new util.ArrayList[E]
    val minX: Int = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D)
    val maxX: Int = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D)
    val minZ: Int = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D)
    val maxZ: Int = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D)

    //world.getEntitiesWithinAABB()
    for (i1 <- minX to maxX)
    {
      for (j1 <- minZ to maxZ)
      {
        if (world.getChunkProvider.chunkExists(i1, j1))
        {
          val chunk = world.getChunkFromChunkCoords(i1, j1)

          var i = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D);
          var j = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D);
          i = MathHelper.clamp_int(i, 0, chunk.entityLists.length - 1);
          j = MathHelper.clamp_int(j, 0, chunk.entityLists.length - 1);

          for (k <- i to j)
          {
            val list1 = chunk.entityLists(k);

            for (l <- 0 to list1.size())
            {
              val entity = list1.get(l).asInstanceOf[Entity];

              if (ClassIsAssignableFrom.isAssignableFrom(clazz, entity.getClass()) && distance3D(new Vector3(entity)) <= r)
              {
                list.add(entity.asInstanceOf[E]);
              }
            }
          }
        }
      }
    }
    return list
  }

}

File: src/main/scala/com/builtbroken/mc/lib/transform/region/Sphere.java
package com.builtbroken.mc.lib.transform.region;

import com.builtbroken.mc.lib.transform.vector.Pos;
import net.minecraft.entity.Entity;
import net.minecraft.util.MathHelper;
import net.minecraft.world.World;
import net.minecraft.world.chunk.Chunk;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by robert on 1/12/2015.
 */
public class Sphere extends Shape3D
{
    double r;

    public Sphere(Pos pos, int r)
    {
        super(pos);
        this.r = r;
    }

    @Override
    public double getSizeX() { return r * 2; }

    @Override
    public double getSizeY() { return r * 2; }

    @Override
    public double getSizeZ() { return r * 2; }

    @Override
    public double getArea() { return 4 * Math.PI * (r * r); }

    @Override
    public double getVolume() { return (4 * Math.PI * (r * r * r)) / 3; }

    public boolean isWithin(double x, double y, double z)
    {
        return new Pos(x, y, z).subtract(x, y, z).magnitude() <= this.r;
    }

    public <E extends Entity> List<E> getEntities(World world, Class<E> clazz)
    {
        List<E> list = new ArrayList();
        int minX = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D);
        int maxX = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D);
        int minZ = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D);
        int maxZ = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D);

        //world.getEntitiesWithinAABB()
        for (int i1 = minX; i1 <= maxX; i1++)
        {
            for (int j1 = minZ; j1 <= maxZ; j1++)
            {
                if (world.getChunkProvider().chunkExists(i1, j1))
                {
                    Chunk chunk = world.getChunkFromChunkCoords(i1, j1);

                    int i = MathHelper.floor_double((r - World.MAX_ENTITY_RADIUS) / 16.0D);
                    int j = MathHelper.floor_double((r + World.MAX_ENTITY_RADIUS) / 16.0D);
                    i = MathHelper.clamp_int(i, 0, chunk.entityLists.length - 1);
                    j = MathHelper.clamp_int(j, 0, chunk.entityLists.length - 1);

                    for (int k = i; k <= j; k++)
                    {
                        List list1 = chunk.entityLists[k];

                        for (int l = 0; l < list1.size(); l++)
                        {
                            Entity entity = (Entity) list1.get(l);

                            if (ClassIsAssignableFrom.isAssignableFrom(clazz, entity.getClass()) && distance(new Pos(entity)) <= r)
                            {
                                list.add((E)entity);
                            }
                        }
                    }
                }
            }
        }
        return list;
    }
}
